### PR TITLE
i2271: Refactor of the montagu-machine code

### PR DESCRIPTION
**Pull request by richfitz**
This moves the provisioning code (and therefore necessarily the
Vagrantfile) to the top level of the repo, so that we can use
these scripts to provision physical machines as well as virtual.

The systemctl bits move into their own directory but remain essentially
unchanged

Some renames are simply dropping '.sh' extensions

Alternative approach, but same content, as #1

### CODE REVIEW

File: Vagrantfile [hunk 1]
```diff
@@ -7,17 +7,26 @@ box = "bento/ubuntu-16.04"
 ram_size_mb = '16384'
 data_disk_size_gb = 900
 
+vault_config_file = 'staging/shared/vault_config'
```
Comment (richfitz) [inline Vagrantfile:10]: This set of lines allows using vault during provisioning

File: Vagrantfile [hunk 2]
```diff
@@ -43,13 +52,24 @@ Vagrant.configure(2) do |config|
   config.persistent_storage.size = data_disk_size_gb * 1024
 
   config.vm.provision :shell do |shell|
-    shell.path = 'provision/setup-docker.sh'
+    shell.path = 'provision/setup-docker'
```
Comment (richfitz) [inline Vagrantfile:55]: The setup-docker script is a bit more flexible now - see the script for details.  We can modify instructions elsewhere to use that script for setting up physical machines

File: Vagrantfile [hunk 3]
```diff
@@ -43,13 +52,24 @@ Vagrant.configure(2) do |config|
...
   end
   config.vm.provision :shell do |shell|
-    shell.path = 'provision/setup-vault.sh'
+    shell.path = 'provision/setup-vault'
+  end
+  config.vm.provision :shell do |shell|
+    shell.path = 'provision/setup-machine-metrics'
+  end
+
+  config.vm.provision :shell do |shell|
```
Comment (richfitz) [inline Vagrantfile]: This step was always manual previously

File: provision/setup-automatic-updates [hunk 4]
```diff
@@ -5,7 +5,7 @@ set -e
 apt-get update
 apt-get install -y unattended-upgrades curl
 
-tee /etc/apt/apt.conf.d/20auto-upgrades <<EOF > /dev/null
+cat <<EOF > /etc/apt/apt.conf.d/20auto-upgrades
```
Comment (richfitz) [inline provision/setup-automatic-updates:8]: This file just swaps `tee` for the more usual `cat`. This was my fault initially in not giving clear instructions

File: provision/setup-docker [hunk 5]
```diff
@@ -1,16 +1,34 @@
 #!/usr/bin/env bash
-set -x
+set -ex
```
Comment (richfitz) [inline provision/setup-docker:2]: More flexible version of the old script

- environment variable `DOCKER_LARGE_DISK` is used to indicate where to put docker's large files (we do this on the annex and I have it this way locally)
- all other arguments are users that will be granted permission to use the docker socket

File: provision/setup-machine-metrics [hunk 6]
```diff
@@ -0,0 +1,18 @@
+#!/usr/bin/env bash
```
Comment (richfitz) [inline provision/setup-machine-metrics:1]: This seems worth testing. If we ever want to monitor these vms this way we'll need to expose the ports via the Vagrantfile but point them at different numbers...

File: provision/setup-montagu [hunk 7]
```diff
@@ -0,0 +1,60 @@
+#!/usr/bin/env bash
```
Comment (richfitz) [inline provision/setup-montagu:1]: This is based off of an restore script that was never really used. But with the vault setup now included we can just go ahead and set everything up

File: provision/setup-montagu [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+    echo "setting up montagu-data"
+    git clone git@github.com:vimc/montagu-data.git $MONTAGU_DATA_PATH
+fi
+
+if [ -d montagu ]; then
+    echo "montagu already set up"
+else
+    echo "setting up montagu"
+    git clone --recursive https://github.com/vimc/montagu
+    git -C ${MONTAGU_PATH} checkout i1472
```
Comment (richfitz) [inline provision/setup-montagu]: This line needs removing once https://github.com/vimc/montagu/pull/80 is merged

File: provision/setup-vagrant [hunk 9]
```diff
@@ -0,0 +1,38 @@
+#!/usr/bin/env bash
```
Comment (richfitz) [inline provision/setup-vagrant:1]: This is designed for running on the host machine, e.g. in setting up `support` again

File: staging/scripts/destroy-disk [hunk 10]
```diff
@@ -0,0 +1,16 @@
+#!/usr/bin/env bash
```
Comment (richfitz) [inline staging/scripts/destroy-disk:1]: This is meant to help with deleting disks without remembering or copying/pasting weird commands

File: staging/scripts/setup-start-on-boot [hunk 11]
```diff
@@ -0,0 +1,22 @@
+#!/usr/bin/env bash
```
Comment (richfitz) [inline staging/scripts/setup-start-on-boot:1]: see staging/scripts/start-on-boot.sh - mostly unchanged

File: staging/scripts/vault-prepare [hunk 12]
```diff
@@ -1,9 +1,15 @@
 #!/usr/bin/env bash
+set -e
```
Comment (richfitz) [inline staging/scripts/vault-prepare:2]: This script is now made a little more robust with respect to where it is run from

File: staging/scripts/write-ssh-shortcuts [hunk 13]
```diff
@@ -0,0 +1,13 @@
+#!/usr/bin/env bash
```
Comment (richfitz) [inline staging/scripts/write-ssh-shortcuts:1]: For use on support for writing out the ssh shortcuts

**Review by richfitz — COMMENTED** [review]
comments explaining some bits added

File: provision/setup-docker [hunk 14]
```diff
@@ -1,16 +1,34 @@
...
+COMPOSE_VERSION=1.22.0
+
+if [ ! -z $DOCKER_LARGE_DISK ]; then
+    if [ ! -d $DOCKER_LARGE_DISK ]; then
+        echo "Expected a directory at $DOCKER_LARGE_DISK"
+        exit 1
+    fi
+
+    VAR_LIB_DOCKER="${DOCKER_LARGE_DISK}/docker/var-lib-docker"
+    if [ -d $DOCKER_LARGE_DISK/docker/var-lib-docker ]; then
```
Comment (hillalex) [inline provision/setup-docker]: can use the `VAR_LIB_DOCKER` assigned on the previous line here

File: provision/setup-docker [hunk 15]
```diff
@@ -1,16 +1,34 @@
...
+        echo "Docker large disk already configured"
+        exit 0
```
Comment (hillalex) [inline provision/setup-docker]: Should we be exiting at this point? Don't we still have to make sure `/var/lib/docker` points to the right place and that docker is installed?

**Review by hillalex — COMMENTED** [review]
Looks good, but lets talk through in person to make sure I understand what's going on with the provisioning logic

File: provision/setup-docker (hunk 15 above)
Comment (richfitz) [reply provision/setup-docker]: yes!

**Comment by richfitz** [comment]
Reminder for me: Depends on https://github.com/vimc/machine-metrics/pull/1/files and  https://github.com/vimc/montagu/pull/80 being merged, at which point the checkouts in the above can be removed

File: provision/setup-vault [hunk 16]
```diff
@@ -0,0 +1,18 @@
+#!/usr/bin/env bash
+set -ex
+
+vault_version=0.11.2
+
+if which -a vault > /dev/null; then
+    echo "vault is already installed"
```
Comment (hillalex) [inline provision/setup-vault:7]: we could probably always run this, as it's pretty quick and means you always get the right version of vault installed?

File: provision/setup-montagu [hunk 17]
```diff
@@ -0,0 +1,62 @@
...
+else
+    echo "setting up montagu-data"
+    git clone git@github.com:vimc/montagu-data.git $MONTAGU_DATA_PATH
+fi
+
+if [ -d montagu ]; then
+    echo "montagu already set up"
+else
+    echo "setting up montagu"
+    git clone --recursive https://github.com/vimc/montagu
```
Comment (hillalex) [inline provision/setup-montagu]: need to specify $MONTAGU_PATH here

**Review by hillalex — CHANGES_REQUESTED** [review]
Couple of small things I noticed when running locally. Also are the steps to set up production documented anywhere?

**Comment by richfitz** [comment]
The only docs we have are the ones in `docs` itself - there's a `Production.md` file there but it could be expanded greatly!

File: provision/setup-vault (hunk 16 above)
Comment (richfitz) [reply provision/setup-vault:7]: ok, the new version installs vault more enthusiastically now

**Review by hillalex — APPROVED** [review]

**Pull request merged**